### ISSUE TITLE
[Bug]: Nextcloud cannot create previews for jpg images 34 MP or larger

### ISSUE THREAD

**Issue by Slartibartfast27** [nextcloud/server#31284]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [x] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Nextcloud cannot create previews for jpg images 34 MP or larger.

### Steps to reproduce

1. Start a fresh docker container with any of these nextcloud versions: 23.0.2 or  or  22.2.5   or  21.0.9 
2. Go to files and drop the following images (recommended: one by one)
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/30_MP_-_JPG-Quality_80.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/32_MP_-_JPG-Quality_80.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/34_MP_-_JPG-Quality_10.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/34_MP_-_JPG-Quality_80.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/36_MP_-_JPG-Quality_80.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/38_MP_-_JPG-Quality_80.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/40_MP_-_JPG-Quality_80.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/45_MP_-_JPG-Quality_1.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/45_MP_-_JPG-Quality_50.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/45_MP_-_JPG-Quality_80.jpg
   * https://files.bulitta.com/dynsrc/2022-02-20-nextcloud-bug-large-images/45_MP_-_JPG-Quality_90.jpg
3. Wait after each dropped picture for the CPU to become idle
4. Refresh the browser
5. Now you should get previews (thumbnails + larger previews) for all the dropped pictures, but actually you only get previews for pictures smaller than 34 MP, no matter how many bytes the jpg file is: 
![2022-02-20_19_03_46-Photos_-_Dateien_-_Nextcloud](https://user-images.githubusercontent.com/7062914/154858924-a697761a-3efb-4b73-89d2-ea5c72de61cc.png)
![2022-02-20_19_04_44-34_MP_-_JPG-Quality_10 jpg_-_Nextcloud](https://user-images.githubusercontent.com/7062914/154858930-2d25c5ff-a13d-4221-9b97-bfae959afa95.png)

### Expected behavior

Nextcloud can create previews for ALL jpg images, even 34 MP or larger.

### Installation method

Official Docker image

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.0

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
root@odroid-DMZ-webserver:/mnt/rootfs-sd/docker/webserver/mountpoints/var-www-html/sd-card# docker exec -it -u www-data nextcloud-test php occ config:list system
{
    "system": {
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "overwritehost": "pubcloud-test.bulitta.com",
        "overwriteprotocol": "http",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "pubcloud-test.bulitta.com",
            "nextcloud"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "23.0.2.1",
        "overwrite.cli.url": "http:\/\/localhost",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***"
    }
}
```

### List of activated Apps

```shell
root@odroid-DMZ-webserver:/mnt/rootfs-sd/docker/webserver/mountpoints/var-www-html/sd-card# docker exec -it -u www-data nextcloud-test php occ app:list
Enabled:
  - accessibility: 1.9.0
  - activity: 2.15.0
  - circles: 23.0.1
  - cloud_federation_api: 1.6.0
  - comments: 1.13.0
  - contactsinteraction: 1.4.0
  - dashboard: 7.3.0
  - dav: 1.21.0
  - federatedfilesharing: 1.13.0
  - federation: 1.13.0
  - files: 1.18.0
  - files_pdfviewer: 2.4.0
  - files_rightclick: 1.2.0
  - files_sharing: 1.15.0
  - files_trashbin: 1.13.0
  - files_versions: 1.16.0
  - files_videoplayer: 1.12.0
  - firstrunwizard: 2.12.0
  - logreader: 2.8.0
  - lookup_server_connector: 1.11.0
  - nextcloud_announcements: 1.12.0
  - notifications: 2.11.1
  - oauth2: 1.11.0
  - password_policy: 1.13.0
  - photos: 1.5.0
  - privacy: 1.7.0
  - provisioning_api: 1.13.0
  - recommendations: 1.2.0
  - serverinfo: 1.13.0
  - settings: 1.5.0
  - sharebymail: 1.13.0
  - support: 1.6.0
  - survey_client: 1.11.0
  - systemtags: 1.13.0
  - text: 3.4.0
  - theming: 1.14.0
  - twofactor_backupcodes: 1.12.0
  - updatenotification: 1.13.0
  - user_status: 1.3.1
  - viewer: 1.7.0
  - weather_status: 1.3.0
  - workflowengine: 2.5.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - user_ldap
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
{"reqId":"FQqMvdjwHvM3de3F3r2o","level":3,"time":"2022-02-20T18:01:23+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=221&c=01e189d81ed2a8337fc740390119010e&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"jv5ECJZaMVGWFjNrvk6s","level":3,"time":"2022-02-20T18:01:36+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=232&c=13f7ef88c690455ed569ed80b2b22149&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"NMtE1ulnGdqPE01frb6p","level":3,"time":"2022-02-20T18:01:46+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=243&c=585af979e49838250c46560643d76917&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"u79kH1hK81nxLIRgFMYJ","level":3,"time":"2022-02-20T18:01:54+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=254&c=9e65626e725cfd1ca934904f1cdbce01&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"rw0n8jaKYU599BWvY4WO","level":3,"time":"2022-02-20T18:01:58+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=262&c=47e65e01e2c7296efe05301a5c9e0d23&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"xxaPzD3STgBzmgIwUL0i","level":3,"time":"2022-02-20T18:02:04+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=270&c=e057c1c9763a77f3fa0f9fa0a87438b7&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"KEYJyBNjDv4FKnUtcacA","level":3,"time":"2022-02-20T18:02:11+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=281&c=e0becdf3db23a5298aad35f08f73377a&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"LOs11UREghcw3UFaAq2h","level":3,"time":"2022-02-20T18:02:18+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=294&c=d596b85c4b0bb2ea6b462f80485afee5&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"v3j1UTlcCiFfdJgcKq0K","level":3,"time":"2022-02-20T18:03:01+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=302&c=887a09279d92916fa2bcb42ba002ef8a&x=313&y=313&forceIcon=0&a=0","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"egpvqvm1klWc35wFaPTq","level":3,"time":"2022-02-20T18:04:38+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=221&x=4096&y=2160&a=true","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"Wk320Ya5KyJECV6EHLEv","level":3,"time":"2022-02-20T18:04:38+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=302&x=4096&y=2160&a=true","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"slsCw5TkhfsCFJjz3VaY","level":3,"time":"2022-02-20T18:04:38+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=232&x=4096&y=2160&a=true","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"8vtlf485QD61cDLFlvf2","level":3,"time":"2022-02-20T18:04:41+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=302&x=4096&y=2160&a=true","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
{"reqId":"45Q4RwXQtgLdn684C35H","level":3,"time":"2022-02-20T18:04:41+00:00","remoteAddr":"172.29.0.3","user":"admin","app":"core","method":"GET","url":"/core/preview?fileId=221&x=4096&y=2160&a=true","message":"OC_Image::fixOrientation(): No image loaded","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0","version":"23.0.2.1"}
```

### Additional info

I already tried to figure out the differences between the last working version 20.0.14 , the first not working version 21.0.9 and the currently latest version 23.0.2 using phpinfo.

The candidate, which looks must interesting for me is the version change of the **imagick module** :

|                            | nextcloud:20 | nextcloud:21 | nextcloud:23 |
| -------------------------- | ------------ | ------------ | ------------ |
| Previews                   | OK           | FAIL         | FAIL         |
| PHP Version                | 7.4.26       | 7.4.28       | 8.0.16       |
| **imagick module version** | **3.5.1**    | **3.7.0**    | **3.7.0**    |
| memory_limit               | 512M         | 512M         | 512M         |

As I am not a php professional, I tried out only a few simple things like this:

```php
<?php
/* Read the image */
#$im = new Imagick("/var/www/html/data/admin/files/Photos/2020-07-20_12-26-29-_U2A6017-A.jpg"); // small
$im = new Imagick("/var/www/html/data/admin/files/Photos/2021-08-08_18-14-44-_N6A2872-A.jpg");  // large

/* Thumbnail the image */
$im->thumbnailImage(200, null);

/* Output the image*/
$im->writeImages( 'thumbnail.jpg', true );
?>
```

Unfortunately imagick worked for ALL pictures in this test snippet code.

Any ideas how to fix this?

**Comment by Slartibartfast27** [comment]
Another thing I already tried without success:
Increase the php memory by setting this environment variable in docker:
`PHP_MEMORY_LIMIT=1024M`

**Comment by acsfer** [comment]
Did you review all these params?

https://github.com/nextcloud/server/blob/master/config/config.sample.php#L1033-L1086

**Comment by Slartibartfast27** [comment]
Dear @acsfer , thanks to much for the precise hint.
Changing `preview_max_memory' => 256` did excactly what I needed.

Do you eventually have an additional hint what to do after doing an "rm -rf" of the previews-folder in order to regenerate all previews in my productive instance? Currently I'm doing a 
```
rm -rf /var/www/html/data/appdata_ocou40f4ox3v/preview/*
add a suffix to all jpg files
docker exec -u www-data nextcloud ./occ files:scan --all 
delete suffix from all jpg files
docker exec -u www-data nextcloud ./occ files:scan --all 
time docker exec -u www-data nextcloud ./occ  preview:generate-all -vvv
```

**Comment by moviuro** [comment]
This issue is still present (last check mid-March on my instance).

**Comment by MaelAbgrall** [comment]
Hello, I highly discourage running the commands from @Slartibartfast27, I'm still figuring out why, but it deleted 10 years of pictures

**Comment by Slartibartfast27** [comment]
> This issue is still present (last check mid-March on my instance).

Hello @moviuro ,
have you tried  changing `'preview_max_memory' => 256` ?
For me that solved the problem.

It it also works for you, I recommend to the nextcloud team to make `'preview_max_memory' => 256` the default value as image sizes are nowadays often more than 34 MP.

**Comment by Lipown** [comment]
Running in Docker in Proxmox LXC. Cointaner has 8GB ram. After setting `'preview_max_memory' => 256 ` and going to photo folders, the OMM killer is executed. Any solution to this?
PHP memory limit is set to 3GB.

**Comment by Slartibartfast27** [comment]
Hello @Lipown , I don't have a solution in your case. As the original author of this issue I just wanted to add my hardware details.
I'm using an (very old but 64 bit) Odroid C2 ( https://www.hardkernel.com/shop/odroid-c2/ ), which has 2 GB RAM + 4 GB swap configured. OS is Armbian 22.02 Jammy  ( https://www.armbian.com/odroid-c2/ ), basically an ubuntu 22.02.
```
uname -a
Linux odroid-DMZ-webserver 5.15.27-meson64 #trunk.0005 SMP PREEMPT Wed Mar 9 10:19:52 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
```

**Comment by moviuro** [comment]
> > This issue is still present (last check mid-March on my instance).
> 
> Hello @moviuro , have you tried changing `'preview_max_memory' => 256` ? For me that solved the problem.
> 
> It it also works for you, I recommend to the nextcloud team to make `'preview_max_memory' => 256` the default value as image sizes are nowadays often more than 34 MP.

Hi, no I have not tried changing this setting. It feels like just hiding it under the rug. That change will probably work until the new standard is 800MP (or whatever). I'm more interested in a definitive solution (such as downscaling the image first, then generating the image which should be functional with the least amount of required memory -- as explained in Nextcloud's docs for minimum system requirements).

**Comment by gmeks** [comment]
Using higher value of preview_max_memory fixed this for me also, i would also recommend increasing the value. And with a bonus of it failing gracefully, so when someone tries to open a to large image, perhaps the raw image should be sendt to the client instead of failture.

**Comment by darmbrust** [comment]
Is there a trick to get it to regenerate thumbnails after making this change?  I'm still seeing the issue after bumping the memory params.

**Comment by jmbluethner** [comment]
TL;DR for this whole conversation for anyone else stumbling upon this.

- Set the Parameter <code>'preview_max_memory' => 256</code> in <code>PATH_TO_NEXTCLOUD/config/config.php</code> so previews for larger photos will be generated
- Optional recommendation from my side: Install the Plugin/App <a href="https://apps.nextcloud.com/apps/previewgenerator">Preview Generator</a>  

My personal opinion to this is that I think the <code>preview_max_memory</code> parameter should be something Nextcloud warns you about on the <code>/settings/admin/overview</code> page. The problem here is that this isn't something that appears in the docs. Also, Nextcloud doesn't tell you that this might be an issue. In the Settings overview you can see a lot of hints and warnings, for example when your cron jobs aren't running as they should, but you can't see anywhere that your previews are not being generated because of the memory limit. I've been searching for two hours before finding this issue which than helped me to solve this problem.

Btw. I'm using Nextcloud 23.0.3 through the official Docker image.

Cheers.

**Comment by hul51u** [comment]
> TL;DR for this whole conversation for anyone else stumbling upon this.
> 
>     * Set the Parameter `'preview_max_memory' => 256` in `PATH_TO_NEXTCLOUD/config/config.php` so previews for larger photos will be generated
> 
>     * Optional recommendation from my side: Install the Plugin/App [Preview Generator](https://apps.nextcloud.com/apps/previewgenerator)
> 
> 
> My personal opinion to this is that I think the `preview_max_memory` parameter should be something Nextcloud warns you about on the `/settings/admin/overview` page. The problem here is that this isn't something that appears in the docs. Also, Nextcloud doesn't tell you that this might be an issue. In the Settings overview you can see a lot of hints and warnings, for example when your cron jobs aren't running as they should, but you can't see anywhere that your previews are not being generated because of the memory limit. I've been searching for two hours before finding this issue which than helped me to solve this problem.
> 
> Btw. I'm using Nextcloud 23.0.3 through the official Docker image.
> 
> Cheers.

'preview_max_memory' => 256 works for me. thanks!

**Comment by jmbluethner** [comment]
@hul51u No worries, ur welcome ✌

**Comment by psijpkes** [comment]
I have the exact same issue at the moment. So I like to set 'preview_max_memory', but unfortunately I cannot find where to set this variable.  I know it is stated `PATH_TO_NEXTCLOUD/config/config.php`, but also this file I  cannot find. I used grep and find, but unsuccesful.  

I make use of Truecharts

**Comment by hul51u** [comment]
You probably have to add this variable.

**Comment by psijpkes** [comment]
> You probably have to add this variable.

Yes I know, the problem is I don't know where to set it. I searched every thing and cannot find the config directory to add ot change the config.php

**Comment by neoteknic** [comment]
Same problem, fixed with 'preview_max_memory' => 1024,  (run without dock on rpi 8Go with php 6G limit), it should be by default somehintg like this :  'preview_max_memory' => PHP_MEMORY_LIMIT

**Comment by pierreozoux** [comment]
A user just reported me this.

I realized that the image is 8000 pixels large.

I understand that there are some limits for the system to generate previews.
(It would be great if it would always be possible to generate previews, this would be the perfect solution, but I understand that this is not always possible)

Then the question is how to communicate to the end user. A better communication would be:
"Sorry, your image is too large to get a preview, resize your image or ask your admin to change this setting"
(But I also understand that coding a message for each error is complicated :upside_down_face: )

Thanks for all the help in this thread, in my case, I'll increase my limits :)

**Comment by Goldenflamer** [comment]
This workaround also works on Nextcloud 29 why the issue appears again

**Comment by maxi1134** [comment]
I have my php memory limit set at 10gb and my preview memory limit at 8gb.

Yet I am unable to preview 200MP pictures in HEIC
